Add CombineAuthorizeFilters option to enable intended authorize filter combinations

HaoK opened the pull request:
Fixes https://github.com/aspnet/Mvc/issues/6995

- Opt in additive behavior, should be a non breaking change
- new mvcOptions.CombineAuthorizeFilters = true to control behavior, new properties on ResourceInvoker/AuthorizationFilterContext to pass the setting around
- If set, the first AuthorizeFilter will combine all authorize filters into a single effective AuthorizationPolicy and use that, while all other filters will be come no-ops

@rynowak @Tratcher @javiercn @blowdart

pranavkm commented on src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvokerProvider.cs:
```diff
@@ -32,6 +34,7 @@ public class ControllerActionInvokerProvider : IActionInvokerProvider
             _maxModelValidationErrors = optionsAccessor.Value.MaxModelValidationErrors;
             _logger = loggerFactory.CreateLogger<ControllerActionInvoker>();
             _diagnosticSource = diagnosticSource;
+            _combineAuthorizeFilters = optionsAccessor.Value.CombineAuthorizeFilters;
```
Instead of plumbing this value through, could we use an ApplicationModelProvider and have it produce an effective `AuthorizationFilter` that does this? Am I missing some a scenario where this approach wouldn't work?

rynowak commented on src/Microsoft.AspNetCore.Mvc.Abstractions/Filters/AuthorizationFilterContext.cs:
```diff
@@ -19,14 +19,34 @@ public class AuthorizationFilterContext : FilterContext
+        public AuthorizationFilterContext(
+            ActionContext actionContext,
+            IList<IFilterMetadata> filters,
+            bool combineAuthorizeFilters)
```
This isn't the right level for an option like this to appear.

rynowak replied:
What I mean to say is that this is much too visible of a place for this to show up, things in this package live forever.

rynowak commented on src/Microsoft.AspNetCore.Mvc.Core/Authorization/AuthorizeFilter.cs:
```diff
@@ -113,6 +113,20 @@ public AuthorizeFilter(string policy)
             }
 
             var effectivePolicy = Policy;
+            if (context.CombineAuthorizeFilters)
```
This should be part of MvcOptions and global for the app, either you are aware of this behavior and want it, or things should work as is.

rynowak replied:
If all of the new behavior is in here, then this class can just read options

HaoK replied:
Ok that's easier, so just service locate the MvcOptions from HttpContext and then I can get rid of plumbing the options everywhere?

rynowak replied:
Yes, that would be better

HaoK commented:
Updated, changes now only in AuthorizeFilter and MvcOptions

pranavkm commented on src/Microsoft.AspNetCore.Mvc.Core/Authorization/AuthorizeFilter.cs:
```diff
@@ -113,6 +114,22 @@ public AuthorizeFilter(string policy)
+            if (mvcOptions.CombineAuthorizeFilters)
+            {
+                // Combine all authorize filters into single effective policy that's only run on the first filter
+                if (context.Filters.First(f => f is AuthorizeFilter) != this)
```
`if (!context.IsEffectivePolicy<AuthorizeFilter>(this))`

rynowak replied:
Yeah, use that method, You actually need to check for last instead of first.

pranavkm commented on src/Microsoft.AspNetCore.Mvc.Core/Authorization/AuthorizeFilter.cs:
```diff
@@ -113,6 +114,22 @@ public AuthorizeFilter(string policy)
+
+                var authorizeFilters = context.Filters.OfType<AuthorizeFilter>();
+                if (authorizeFilters.Count() > 1) {
+                    effectivePolicy = authorizeFilters.Select(s => s.Policy).Aggregate((p1, p2) => AuthorizationPolicy.Combine(p1, p2));
```
```C#
AuthorizationPolicyBuilder builder;
for (var i = 0; i < context.Filters.Count; i++)
{
    if (ReferenceEquals(this, context.Filters[i]))
    {
        continue;
    }
    
    if (context.Filters[i] is AuthorizeFilter authorizeFilter)
    {
        builder = builder ?? new AuthorizationPolicyBuilder(effectivePolicy);
        builder.Combine(authorizeFilter.Policy);
    }
}

effectivePolicy = builder?.Build() ? effectivePolicy;
```

pranavkm replied:
That said, if this is a common-ish scenario, doing this in a filter created via ApplicationModel would avoid the allocations here entirely. Just something to consider.

rynowak replied:
Is there any way we can cache this? I'm not stoked on the idea of all of this happening on each request

HaoK replied:
We can cache if the policy provider is null just like the other policy code

HaoK replied:
Related to this, I should add some custom policy provider tests to this to make sure this isn't horribly broken when using those

javiercn replied:
Yeah, we should do per request work only if the user uses a custom IAuthorizationPolicyProvider

pranavkm commented on src/Microsoft.AspNetCore.Mvc.Core/MvcOptions.cs:
```diff
@@ -185,5 +185,11 @@ public int MaxModelValidationErrors
         public bool SuppressJsonDeserializationExceptionMessagesInModelState { get; set; } = false;
+
+        /// <summary>
+        /// Gets or sets a flag to determine whether AuthorizeFilters will be combined into a single
```
> Gets or sets a value that determines if policies on instances of `<see cref="AuthorizeFilter" />`...

pranavkm commented on src/Microsoft.AspNetCore.Mvc.Core/MvcOptions.cs:
```diff
@@ -185,5 +185,11 @@ public int MaxModelValidationErrors
+        /// Gets or sets a flag to determine whether AuthorizeFilters will be combined into a single
+        /// effective policy. This was always to be the intended behavior, but was not the case.
+        /// </summary>
+        public bool CombineAuthorizeFilters { get; set;}
```
Space after `set;`

HaoK commented on src/Microsoft.AspNetCore.Mvc.Core/MvcOptions.cs:
```diff
@@ -185,5 +185,11 @@ public int MaxModelValidationErrors
+
+        /// <summary>
+        /// Gets or sets a flag to determine whether AuthorizeFilters will be combined into a single
+        /// effective policy. This was always to be the intended behavior, but was not the case.
```
Well i wasn't sure how to describe why you would turn this on :)

HaoK commented on src/Microsoft.AspNetCore.Mvc.Core/Authorization/AuthorizeFilter.cs:
```diff
@@ -113,6 +114,22 @@ public AuthorizeFilter(string policy)
+            var mvcOptions = context.HttpContext.RequestServices.GetRequiredService<IOptions<MvcOptions>>().Value;
+            if (mvcOptions.CombineAuthorizeFilters)
+            {
+                // Combine all authorize filters into single effective policy that's only run on the first filter
```
Is that what `if (!context.IsEffectivePolicy<AuthorizeFilter>(this))` does?

rynowak replied:
Yes, that chooses the closest filter

HaoK commented on test/WebSites/FiltersWebSite/Startup.cs:
```diff
@@ -16,10 +16,11 @@ public class Startup
                     .AddScheme<BasicOptions, BasicAuthenticationHandler>("Interactive", _ => { })
                     .AddScheme<BasicOptions, BasicAuthenticationHandler>("Api", _ => { });
-            services.AddMvc();
+            services.AddMvc(o => o.CombineAuthorizeFilters = true);
```
Well, its a fairly subtle difference in behavior, the existing functional tests had no change with this setting by default, but we do have coverage in the existing unit tests as those failed with the Combine flag set...

I had to add the new logic to the tests that verify the new functionality works (via the counts on the principals).  The most important test for this feature is actually the new `CanReachActionWhenControllerAuthorizeFails` test since that's how many people stumbled into this issue where they have a parent `[Authorize]` (that requires cookies), and a `[Authorize("Bearer"]` on the action, and even when they had a Bearer token, the request would get redirected to the login page since there were two distinct filters being run.

rynowak replied:
I think the point that @kichalla is making is that it's possible to add new startup classes to our existing projects so that you can test both cases if you like, and more importantly without impacting any other tests.

rynowak replied:
We don't really like to see changes to global configuration because it's not clear what the tests expectations were. A case that we were testing might no longer be tested.

HaoK replied:
Ok yeah I can just add a new set of tests that verify this particular new option/behavior in the non superfund site and leave the existing tests alone then :)

rynowak commented on src/Microsoft.AspNetCore.Mvc.Core/Authorization/AuthorizeFilter.cs:
```diff
@@ -113,6 +114,22 @@ public AuthorizeFilter(string policy)
 
             var effectivePolicy = Policy;
+                            
+            var mvcOptions = context.HttpContext.RequestServices.GetRequiredService<IOptions<MvcOptions>>().Value;
```
Can you add caching for this?

rynowak commented on test/WebSites/FiltersWebSite/Controllers/CombineAuthorizeController.cs:
```diff
@@ -0,0 +1,24 @@
+using Microsoft.AspNetCore.Authorization;
+using System.Linq;
+
+namespace FiltersWebSite
```
Would it be possible to use `SecurityWebSite` for this? `FiltersWebSite` is basically a [superfund](https://en.wikipedia.org/wiki/Superfund) site as far as its usability. I don't want to put any new tests that we care about in this project if at all possible.

HaoK commented:
- Updated with effective policy and mvc options caching
- Moved new combine filter tests into security, the old behavior is tested in the StartupWithGlobalDeny, and a new startup that turns on the option verifies the new behavior

kichalla commented on src/Microsoft.AspNetCore.Mvc.Core/Authorization/AuthorizeFilter.cs:
```diff
@@ -104,15 +108,39 @@ public AuthorizeFilter(string policy)
+                _mvcOptions = context.HttpContext.RequestServices.GetRequiredService<IOptions<MvcOptions>>().Value;
+            }
+
+            if (_mvcOptions.CombineAuthorizeFilters)
```
You seem to have lost the following check in this iteration?
```c#
if (!context.IsEffectivePolicy<AuthorizeFilter>(this)
{
      return;
}
```

HaoK replied:
Looks like it, good catch, I'll see if I can add a test to ensure that only a single filter is calling into the authorize service

HaoK replied:
Ok updated, plugged in a custom PolicyEvaluator that counts calls, and verified that the count was 2 before the fix, and now is 1 as expected.

javiercn commented on src/Microsoft.AspNetCore.Mvc.Core/Authorization/AuthorizeFilter.cs:
```diff
@@ -105,29 +108,37 @@ public AuthorizeFilter(string policy)
+                    {
+                        builder = builder ?? new AuthorizationPolicyBuilder(effectivePolicy);
+                        builder.Combine(authorizeFilter.Policy);
+                    }
```
Why does all this policy computation happen inside the filter instead of in the https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/Internal/AuthorizationApplicationModelProvider.cs

Is there any reason I'm missing here for this not to be the case? /cc: @rynowak 

The only thing I can think of is someone pluging in an authorize filter dynamically through a filter provider but that sounds very far fetched and something we would not recommend our customers.

IMO resolving the effective policy should happen at the application model level and the filter should only make a decision on whether it should execute or not.

javiercn replied:
Ignore this comment. Let's leave this here for now and get the ball rolling, we'll file an issue for this and move it later.

javiercn commented on src/Microsoft.AspNetCore.Mvc.Core/Authorization/AuthorizeFilter.cs:
```diff
@@ -104,15 +108,43 @@ public AuthorizeFilter(string policy)
+
+            if (_mvcOptions.CombineAuthorizeFilters)
+            {
+                if (!context.IsEffectivePolicy<AuthorizeFilter>(this)) {
```
This will cause issues as it will change the order in which the authorization filter executes. I'm going to give a practical example.

I plug in globally (in this order)

```csharp
new AuthorizeFilter(MyPolicy)
new AutoValidateAntiforgeryTokenFilter()
```

I add another Authorize filter in either my controller or my action.

```csharp
[Authorize("2FAPolicy")]
```
Now lets examine the existing and new behaviors:
With the existing behavior
The first authorize filter executes and evaluates the policy (Authenicating the user in the process)
The autovalidate antiforgery filter executes (with the identity provided by 1)
The second authorize filter executes (and sets up identity 2)

With the new behavior
The first authorize filter does nothing (returns).
The autovalidate antiforgery filter executes (with an anonymous identity)
The second authorize filter executes (and sets up a combined identity of 1 and 2)

What is wrong here is that the order in which the AutorizeFilter executes relative to other IAuthorizationFilter changes and that has an implicit impact on the execution of other filters. There are a couple of ways we can go about fixing this (we can do both as they complement each other and have value on their own):
* Instead of making the last filter do the combined authorization, we make the first filter combine the individual policies and execute the filter. (This is correct due to the fact that the semantics here are to combine instead of override).
* Make other IAuthorizationFilters in our platform ordered (Like ValidateAntiforgeryTokenFilter) that way this filter will always execute in the same order relative to those other two. That solves the problem for our built-in filters but not for custom implementations created by our users.

rynowak replied:
Order takes precedence over scope.

This means that AF will run after **ALL** `AuthorizeFilter` instances

javiercn replied:
@rynowak I didn't see https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.ViewFeatures/ValidateAntiForgeryTokenAttribute.cs#L21, so point 2 its covered then.

@rynowak Do we still care about doing
> What is wrong here is that the order in which the AutorizeFilter executes relative to other IAuthorizationFilter changes and that has an implicit impact on the execution of other filters.

This would only affect people building their own filters IAuthorizationFilter implementations that have a dependency on the authorize behavior. (and not ordering them). Do we care about it?

If we don't, then there's no point on making the first authorizefilter execute the combined policy instead of the last. On the other side, it should be cheap to change.

After we resolve this I think this is good to go, given that all our filters (CORS and Antiforgery) are ordered and won't run into any issue with this.

rynowak replied:
> What is wrong here is that the order in which the AutorizeFilter executes relative to other IAuthorizationFilter changes and that has an implicit impact on the execution of other filters.

I would argue that it's right now, and used to be wrong. Anyone who cases about order of execution  relative to AuthorizeFilter needs to use `Order` - **End of story**. We've patched two issues related to that, which is too many 😆 

This will be an opt-in change, so it's not going to take anyone by surprise.

> If we don't, then there's no point on making the first authorizefilter execute the combined policy instead of the last. On the other side, it should be cheap to change.

The last filter should execute the policy. This is how filters work.

HaoK commented:
Yup, Authorize filters that specify authentication schemes have authentication side effects, so running them on the first instance seems to make more sense even if it breaks with the pattern that other filters use.

rynowak commented:
This looks good to me now

pranavkm commented on src/Microsoft.AspNetCore.Mvc.Core/MvcOptions.cs:
```diff
@@ -185,5 +186,12 @@ public int MaxModelValidationErrors
+        /// will be combined into a single effective policy. This was always to be the intended behavior, 
+        /// but was not the case.
+        /// </summary>
+        public bool CombineAuthorizeFilters { get; set; }
```
Work to use CompatSwitch tracked via https://github.com/aspnet/Mvc/issues/7162

HaoK commented:
AppVeyor runs for some reason are very unhappy, with errors that appear to be mismatched dependencies... 

```
System.MissingMethodException : Method not found: 'System.ValueTuple`3<Microsoft.Net.Http.Headers.RangeItemHeaderValue,Int64,Boolean> Microsoft.AspNetCore.Mvc.Infrastructure.FileResultExecutorBase.SetHeadersAndLog(Microsoft.AspNetCore.Mvc.ActionContext, Microsoft.AspNetCore.Mvc.FileResult, System.Nullable`1<Int64>, Boolean, System.Nullable`1<System.DateTimeOffset>, Microsoft.Net.Http.Headers.EntityTagHeaderValue)'.
```

I just rebased to dev yesterday afternoon, is there something else I need to do to get appveyor passing?  When I run locally everything passes except one integration test which fails with a no ObjectPoolProvider service registered error.

rynowak commented:
Rebase again, I pushed a fix for this last night

HaoK commented:
Awesome ok, rebased and I had to add ObjectPoolProvider to the integration tests to make that test pass, I'll merge once appveyor/travis go green

HaoK commented:
73bd09dc1c8abe2ff71cd33679fa17ebd04ff189
